common/src/main/java/io/netty/util/HashedWheelTimer.java

```diff
@@ -337,33 +362,37 @@ public void start() {
 
         if (!WORKER_STATE_UPDATER.compareAndSet(this, WORKER_STATE_STARTED, WORKER_STATE_SHUTDOWN)) {
             // workerState can be 0 or 2 at this moment - let it always be 2.
-            WORKER_STATE_UPDATER.set(this, WORKER_STATE_SHUTDOWN);
-
-            if (leak != null) {
-                boolean closed = leak.close(this);
-                assert closed;
+            if (WORKER_STATE_UPDATER.getAndSet(this, WORKER_STATE_SHUTDOWN) != WORKER_STATE_SHUTDOWN) {
+                INSTANCE_COUNTER.decrementAndGet();
+                if (leak != null) {
+                    boolean closed = leak.close(this);
+                    assert closed;
+                }
             }
 
             return Collections.emptySet();
         }
 
-        boolean interrupted = false;
-        while (workerThread.isAlive()) {
-            workerThread.interrupt();
-            try {
-                workerThread.join(100);
-            } catch (InterruptedException ignored) {
-                interrupted = true;
+        try {
+            boolean interrupted = false;
+            while (workerThread.isAlive()) {
+                workerThread.interrupt();
+                try {
+                    workerThread.join(100);
+                } catch (InterruptedException ignored) {
+                    interrupted = true;
+                }
             }
-        }
 
-        if (interrupted) {
-            Thread.currentThread().interrupt();
-        }
-
-        if (leak != null) {
-            boolean closed = leak.close(this);
-            assert closed;
+            if (interrupted) {
+                Thread.currentThread().interrupt();
+            }
+        } finally {
+            INSTANCE_COUNTER.decrementAndGet();
```

Comment:
it could actually go negative (e.g. overflow) as we don't strictly enforce the limit ... we just use it to warn for "too many instances". In practice I wouldn't expect this to happen but I'm not sure we should add an assert for it unless we strictly enforce it.